compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/tower/TowerResolver.kt

```diff
@@ -186,32 +218,69 @@ class TowerResolver {
                 }
+                implicitScopeTower.getImplicitReceiver(scope)?.let { rv ->
+                    processImplicitReceiver(rv, resolveExtensionsForImplicitReceiver)?.let { return it }
+                }
+
+                val contextReceiversGroup = implicitScopeTower.getContextReceivers(scope)
+                if (contextReceiversGroup.isNotEmpty()) {
+                    contextReceiversGroups.add(contextReceiversGroup)
+                }
                 return null
             }
 
-            if (implicitScopeTower.implicitsResolutionFilter === ImplicitsExtensionsResolutionFilter.Default) {
-                for (scope in implicitScopeTower.lexicalScope.parentsWithSelf) {
-                    processScope(scope, true)?.let { return it }
+            fun processContextReceiverGroup(contextReceiversGroup: List<ReceiverValueWithSmartCastInfo>): Collection<C>? {
+                TowerData.TowerLevel(ContextReceiversGroupScopeTowerLevel(implicitScopeTower, contextReceiversGroup))
+                    .process()?.let { return it }
+                TowerData.BothTowerLevelAndContextReceiversGroup(syntheticLevel, contextReceiversGroup).process()
+                    ?.let { return it }
+                for (nonLocalLevel in nonLocalLevels) {
+                    TowerData.BothTowerLevelAndContextReceiversGroup(nonLocalLevel, contextReceiversGroup).process()
+                        ?.let { return it }
                 }
-            } else {
-                for (scopeInfo in implicitScopeTower.allScopesWithImplicitsResolutionInfo()) {
-                    processScope(scopeInfo.scope, scopeInfo.resolveExtensionsForImplicitReceiver)?.let { return it }
+                return null
+            }
+
+            fun processImportingScope(scope: ImportingScope): Collection<C>? {
+                TowerData.TowerLevel(ImportingScopeBasedTowerLevel(implicitScopeTower, scope))
+                    .process(scope.mayFitForName(name))?.let { return it }
+                return null
+            }
+
+            fun processScopes(
+                scopes: Sequence<HierarchicalScope>,
+                resolveExtensionsForImplicitReceiver: (HierarchicalScope) -> Boolean
+            ): Collection<C>? {
+                var firstImportingScopePassed = false
+                for (scope in scopes) {
+                    if (scope is LexicalScope) {
+                        processLexicalScope(scope, resolveExtensionsForImplicitReceiver(scope))?.let { return it }
+                    } else if (!firstImportingScopePassed) {
+                        firstImportingScopePassed = true
+                        contextReceiversGroups.forEach { contextReceiversGroup ->
+                            processContextReceiverGroup(contextReceiversGroup)?.let { return it }
+                        }
+                    } else {
+                        processImportingScope(scope as ImportingScope)
```

Comment:
It seems that at your branch `FirResolveModularizedTotalKotlinTest.testTotalKotlin` fails when run on Kotlin project, do you have the prepared test data?

Reply:
Thanks, missed it. Fixed